### PR TITLE
Add spec for BroadcastInDimOp

### PR DESCRIPTION
**Pull request by sdasgup3**
Adds the specification as  part of #122 

Fixes https://github.com/openxla/stablehlo/issues/277

### CODE REVIEW

**Review by burmako — CHANGES_REQUESTED** [review]
Thank you for the meticulous work of digging into the MHLO semantics, reconciling it with HLO semantics and then documenting the findings! 

Unless I'm missing something major, I think the documentation can be significantly simplified (search for a comment that starts with "Also, why do we need multiple bullets to describe the semantics"). Please take a look!

File: docs/spec_draft.md [hunk 1]
```diff
@@ -153,6 +153,7 @@ described below)
    * [add](#stablehloadd)
    * [and](#stablehloand)
    * [ceil](#stablehloceil)
+   * [broadcast_in_dim](#stablehlobroadcast_in_dim)
```
Comment (ghpvnist) [inline docs/spec_draft.md:156]: Looks like this is out of order? Same with the actual semantics section.

File: docs/spec_draft.md [hunk 2]
```diff
@@ -362,6 +363,159 @@ IEEE-754 specification.
...
+//            ],
+//            [
+//             [7, 8],
+//             [9, 10],
+//             [11, 12]
+//            ],
+//           ]
+%result = "stablehlo.broadcast_in_dim"(%operand) {
+  broadcast_dimensions = dense<[2, 3, 0]>: tensor<2xi64>
+ } : (tensor<2x1xi32>) -> tensor<2x3x2xi32>
```
Comment (ghpvnist) [inline docs/spec_draft.md:445]: Input tensor shape does not match the `operand` or did I miss something?

File: docs/status.md [hunk 3]
```diff
@@ -41,7 +41,7 @@ one of the following tracking labels.
 | batch_norm_inference     |      no       |     yes*     |       yes*        |        no         |     no      |
 | batch_norm_training      |      no       |     yes*     |       yes*        |        no         |     no      |
 | bitcast_convert          |      no       |     yes*     |    infeasible     |        yes        |     no      |
-| broadcast_in_dim         |      no       |     yes*     |    infeasible     |        no         |     no      |
+| broadcast\_in\_dim       |      yes      |     yes      |    infeasible     |        no         |     no      |
```
Comment (ghpvnist) [inline docs/status.md:44]: I don't think we need the backslash here.

File: stablehlo/dialect/StablehloOps.td [hunk 4]
```diff
@@ -1673,20 +1673,17 @@ def StableHLO_BroadcastInDimOp : StableHLO_Op<"broadcast_in_dim",
       [NoSideEffect, SameOperandsAndResultElementType]> {
   let summary = "Broadcast a tensor into the given shape by adding dimensions.";
   let description = [{
-    Broadcasts the `operand` tensor to a higher rank. This is not the limited
-    form of broadcasting exposed as the XLA client broadcast op, but rather the
-    more powerful "InDim" broadcasting, which is closer to the HLO broadcast op
-    and exposed in the XLA client BroadcastInDim method.
+    Broadcasts the `operand` tensor to a higher rank.
```
Comment (ghpvnist) [inline stablehlo/dialect/StablehloOps.td:1676]: Replace this description with the first sentence of the Semantics section:
```
Expands the dimensions and/or rank of an input tensor by duplicating the data
in the `operand` tensor and produces a `result` tensor.
```

File: stablehlo/dialect/StablehloOps.td [hunk 5]
```diff
@@ -1673,20 +1673,17 @@ def StableHLO_BroadcastInDimOp : StableHLO_Op<"broadcast_in_dim",
...
 
-    `broadcast_dimensions` maps the operand dimension number to the target shape
-    dimension number. It must have the same size as the rank of the operand. The
-    mapped dimensions must either be the same size or the dimension being
-    broadcast from must be size 1 (degenerate broadcasting).
-
-    For a scalar (0D tensor) operand, `broadcast_dimensions` must be empty. The
-    The scalar value will be broadcast to every element in the target shape.
+    See:
+    https://github.com/openxla/stablehlo/blob/main/docs/spec_draft.md#stablehlobroadcastindim
```
Comment (ghpvnist) [inline stablehlo/dialect/StablehloOps.td:1679]: Should be:
`https://github.com/openxla/stablehlo/blob/main/docs/spec_draft.md#stablehlobroadcast_in_dim`

**Review by ghpvnist — CHANGES_REQUESTED** [review]

File: docs/spec_draft.md [hunk 6]
```diff
@@ -362,6 +363,159 @@ IEEE-754 specification.
...
+| Name     | Type                                  |
+|----------|---------------------------------------|
+| `result` | tensor of any supported element types |
+
+### Constraints
+
+  * (C1) `operand` and `result` have the element type.
+  * (C2) size(`broadcast_dimensions`) $=$ rank(`operand`).
+  * (C3) $0 \le$ `broadcast_dimensions`[i] $\lt$ rank(`result`) $\forall
+  i \in$ [0, size(`broadcast_in_dim`)).
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply docs/spec_draft.md:420]: By `size` I mean the number of elements in the 1-D array. Perhaps  `dim(broadcast_in_dim, 0)`

[the remark this answers is not in the document]
Comment (sdasgup3) [reply docs/spec_draft.md:420]: done.

File: docs/spec_draft.md [hunk 7]
```diff
@@ -362,6 +363,159 @@ IEEE-754 specification.
...
+  * (C4) No duplicates in `broadcast_dimensions`.
+  * (C5) For k $\in$ [0, rank(`operand`))
+    * `d`$^{op}$`k` $= 1$ or
+    * `dk` $=$ `d`$^{r}$`j`, where `j` $=$ `broadcast_dimension`[`k`]
+    where `d`$^{r}$`j` is the dimension size of `j`$^{th}$ dimension in `result`
+    and `dk` is the dimension size of `k`$^{th}$ dimension in `operand`.
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply docs/spec_draft.md:426]: the suggestion looks good.

File: docs/spec_draft.md [hunk 8]
```diff
@@ -362,6 +363,159 @@ IEEE-754 specification.
...
+   * Let $P$ = [0, 1, ..., `R`-1], where `R` is the rank of `operand`.
+   * Sort $P$ in the ascending order of `broadcast_dimensions` to get $P'$.
+   * Apply [transpose](#stablehlotranspose) on `operand` using $P'$ as
+  `permutation` to get `transposed_operand`
+   * Sort `broadcast_dimensions` in increasing order of its elements to get
+   `sorted_broadcast_dimension`.
+   * Apply [broadcast_in_dim](#stablehlobroadcast_in_dim) with
+  `transposed_operand` and `sorted_broadcast_dimension`.
+ * For a scalar `operand`, the scalar value will be broadcast to every element
+   in the `result` shape.
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply docs/spec_draft.md:400]: That is right. We do not need it.

File: docs/spec_draft.md [hunk 9]
```diff
@@ -362,6 +363,159 @@ IEEE-754 specification.
...
+in the `operand` tensor and produces a `result` tensor.
+
+Formally, we define the operation as
+  * If `broadcast_dimensions` is sorted in increasing order of dimensions
+    * If rank(`operand`) $=$ rank(`result`):
+      * `result[i0, i1, ..., iR-1]` $=$ `operand[j0, j1, ..., jR-1]` such that
+     `jk` $= 0$ if `dk` $= 1$, where `dk` is the dimension size of `k`$^{th}$
+     dimension in `operand`.
+    * If rank(`operand`) $\ne$ rank(`result`):
+      * [reshape](#stablehloreshape) `operand` to the shape vector $S$,
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply docs/spec_draft.md:380]: The reason I have "vector" is to later say that

... shape vector $S$ of size equals to rank(`result`), defined below,  ..   (I missed saying about "size equals to ..." in the original draft) 

IMO, we need to say what the size is so as to create the S with 1 inserted in places which are not defined by  

 $S$[`broadcast_dimensions`[`k`]] $=$ `dk`, where `dk`
      is the dimension size of `k`$^{th}$ dimension in `operand`, if `k` $\in$
      `broadcast_dimensions


()

File: docs/spec_draft.md (hunk 2 above)
Comment (sdasgup3) [reply docs/spec_draft.md:445]: Yes you are right. Fixed it.

File: docs/spec_draft.md [hunk 10]
```diff
@@ -362,6 +363,159 @@ IEEE-754 specification.
...
+      defined below, to get `reshaped_operand`.
+        * $S$[`broadcast_dimensions`[`k`]] $=$ `dk`, where `dk`
+      is the dimension size of `k`$^{th}$ dimension in `operand`, if `k` $\in$
+      `broadcast_dimensions`
+        * $S$[k] $= 1$, Otherwise.
+      * `result[i0, i1, ..., iR-1]` $=$ `reshaped_operand[j0, j1, ..., jR-1]`
+      such that `jk` $= 0$ if `dk` $= 1$, where `dk` is the dimension size of
+      `k`$^{th}$ dimension in `reshaped_operand`.
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply docs/spec_draft.md]: Thanks! This is much simplified and makes sense.

File: docs/spec_draft.md [hunk 11]
```diff
@@ -326,6 +327,84 @@ logical operation.
...
 [Back to Ops](#index-of-ops)
 
+## stablehlo.broadcast_in_dim
+
+### Semantics
+
+Expands the dimensions and/or rank of an input tensor by duplicating the data
+in the `operand` tensor and produces a `result` tensor. Formally, 
+`result[i0, i1, ..., iR-1]` $=$ `operand[j0, j1, ..., jR-1]` such that
+`jk` $=$ `dim(operand, k) == 1 ? 0 : i[boradcast_dimensions[k]]`, for all
```
Comment (burmako) [inline docs/spec_draft.md:337]: Typo: "broadcast_dimensions".

File: docs/spec_draft.md [hunk 12]
```diff
@@ -326,6 +327,84 @@ logical operation.
...
+### Results
+
+| Name     | Type                                  |
+|----------|---------------------------------------|
+| `result` | tensor of any supported element types |
+
+### Constraints
+
+  * (C1) `operand` and `result` have the element type.
+  * (C2) dim(`broadcast_dimensions`, 0) $=$ rank(`operand`).
```
Comment (burmako) [inline docs/spec_draft.md:356]: `size(broadcast_dimensions) = rank(operand)`? Seems to be a bit shorter than `dim(broadcast_dimensions, 0)`, and is probably frequent enough to warrant having a separate function? E.g. `concatenate`'s (C2) could be rewritten using `size` as well.

File: docs/spec_draft.md [hunk 13]
```diff
@@ -326,6 +327,84 @@ logical operation.
...
+//            [
+//             [2, 4, 6],
+//             [8, 10, 12]
+//            ],
+//            [
+//             [2, 4, 6],
+//             [8, 10, 12]
+//            ],
+//           ]
+//          ]
```
Comment (burmako) [inline docs/spec_draft.md:403]: What do you think about simplifying the example? This pull request took a lot of effort to figure out, but the readers don't have to know about that :) 

E.g. the proposed example below demonstrates broadcasting, as well as the two special cases: 1) non-sorted dimensions, 2) degenerate broadcasting, while involving fewer distinct numbers to track and fewer dimensions.

```mlir
// %operand: [
//            [1, 2, 3],
//           ]
%result = "stablehlo.broadcast_in_dim"(%operand) {
  broadcast_dimensions = dense<[2, 1]>: tensor<2xi64>
} : (tensor<1x3xi32>) -> tensor<2x3x2xi32>
// %result: [
//            [
//             [1, 1],
//             [2, 2],
//             [3, 3]
//            ],
//            [
//             [1, 1],
//             [2, 2],
//             [3, 3]
//            ],
//          ]
```

File: docs/spec_draft.md [hunk 14]
```diff
@@ -1408,7 +1487,9 @@ where `i[d] = j[permutation[d]]`.
   * (C1) `operand` and `result` have the same element type.
   * (C2) `permutation` is a permutation of `[0, 1, ..., R-1]` where `R` is the
   rank of `operand`.
-  * (C3) `result`'s shape is a permutation of `operand`'s shape.
+  * (C3) `result`'s shape is a permutation of `operand`'s shape w.r.t
+  `permutation`. Formally, for all dimensions `i` in `operand`,
+  `dim(operand, i) = dim(result, j)` where `j = permutation[i]`.
```
Comment (burmako) [inline docs/spec_draft.md:1492]: How about we drop the informal part? I.e. only keep "for all dimensions `i` in `operand`,`dim(operand, i) = dim(result, j)` where `j = permutation[i]`".

**Review by burmako — CHANGES_REQUESTED** [review]

File: docs/spec_draft.md (hunk 12 above)
Comment (sdasgup3) [reply docs/spec_draft.md:356]: Agreed about the use of size on 1-D tensors. Added an action task to define these terms  in  https://github.com/openxla/stablehlo/issues/289

**Review by burmako — APPROVED** [review]

File: docs/spec_draft.md [hunk 15]
```diff
@@ -327,6 +328,65 @@ logical operation.
...
+| `result` | tensor of any supported element types |
+
+### Constraints
+
+  * (C1) `operand` and `result` have the element type.
+  * (C2) size(`broadcast_dimensions`) $=$ rank(`operand`).
+  * (C3) $0 \le$ `broadcast_dimensions`[i] $\lt$ rank(`result`) for all
+  dimensions i in `operand`.
+  * (C4) All dimensions in `broadcast_dimensions` are unique.
+  * (C5) For all dimensions `i` in `operand`:
```
Comment (ghpvnist) [inline docs/spec_draft.md:361]: Let's unify `i`, `j`, and `k` to match the variables from the semantics. E.g. `result[i]` is used in semantics section but the same `i` is used in `dim(operand, i)`.

File: stablehlo/dialect/StablehloOps.td [hunk 16]
```diff
@@ -1673,20 +1673,18 @@ def StableHLO_BroadcastInDimOp : StableHLO_Op<"broadcast_in_dim",
...
-
-    For a scalar (0D tensor) operand, `broadcast_dimensions` must be empty. The
-    The scalar value will be broadcast to every element in the target shape.
+    See:
+    https://github.com/openxla/stablehlo/blob/main/docs/spec_draft.md#stablehlobroadcast_in_dim
 
-    See https://www.tensorflow.org/xla/broadcasting.
+    Example:
+    ```mlir
+    "mhlo.broadcast_in_dim"(%operand) {
```
Comment (ghpvnist) [inline stablehlo/dialect/StablehloOps.td:1684]: Change to `stablehlo.broadcast_in_dim`

**Review by ghpvnist — CHANGES_REQUESTED** [review]

File: docs/spec_draft.md [hunk 17]
```diff
@@ -327,6 +328,65 @@ logical operation.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.broadcast_in_dim
+
+### Semantics
+
+Expands the dimensions and/or rank of an input tensor by duplicating the data
+in the `operand` tensor and produces a `result` tensor. Formally,
+`result[i0, i1, ..., iR-1]` $=$ `operand[j0, j1, ..., jR-1]` such that
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply docs/spec_draft.md:337]: This is the same way we define the indices in `Types` section above. IMO, we have shown least two indices to show that the indices are in sequence with step 1. 
If R = 1, then it means we have just i0 as indices, for R = 0, we have no indices.

File: docs/spec_draft.md [hunk 18]
```diff
@@ -327,6 +328,65 @@ logical operation.
...
+
+### Results
+
+| Name     | Type                                  |
+|----------|---------------------------------------|
+| `result` | tensor of any supported element types |
+
+### Constraints
+
+  * (C1) `operand` and `result` have the element type.
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply docs/spec_draft.md]: Great!

File: docs/spec_draft.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (ghpvnist) [reply docs/spec_draft.md:337]: Makes sense, my concern exactly was that by explicitly writing `operand[j0, j1, ..., jR-1]`, that we have at least one dimension in `operand`. Maybe this is something we could bring up in #289 for feedback or decide not to change if its interpretations are clear.

**Comment by sdasgup3** [comment]
incorporated the latest changes #313 and #314

File: docs/spec_draft.md [hunk 19]
```diff
@@ -327,6 +328,65 @@ logical operation.
...
 [Back to Ops](#index-of-ops)
 
+## stablehlo.broadcast_in_dim
+
+### Semantics
+
+Expands the dimensions and/or rank of an input tensor by duplicating the data
+in the `operand` tensor and produces a `result` tensor. Formally,
+`result[i0, i1, ..., iR-1]` $=$ `operand[j0, j1, ..., jR-1]` such that
+`jk` $=$ `dim(operand, k) == 1 ? 0 : i[broadcast_dimensions[k]]`, for all
```
Comment (ghpvnist) [inline docs/spec_draft.md:338]: Nit: extraneous comma
```
`jk` $=$ `dim(operand, k) == 1 ? 0 : i[broadcast_dimensions[k]]` for all
```

**Review by ghpvnist — APPROVED** [review]
LGTM with last round of fixes.

**Pull request merged**